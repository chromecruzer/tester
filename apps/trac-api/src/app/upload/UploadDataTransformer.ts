
import _ from 'lodash';
import util from "util";
import UploadStoreManagement from "../datastore/UploadStoreManagement";
import {TransformData} from "../datastore/TransformData";
import {UploadTransformerA} from "./UploadTransformerA";
import {NotificationsServer} from "@trac/postgresql";
import {InputDataTypeLabel, InputToSqlMapping} from "@trac/datatypes";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export class UploadDataTransformer extends UploadTransformerA {
  private transformData: TransformData;

  constructor(storeFactory: UploadStoreManagement, schemaName: string, tableName,
              private ns = null as (NotificationsServer | null)) {
    super(storeFactory, schemaName, tableName);
    this.transformData = new TransformData(schemaName, tableName, ns);
  }
  async transform(uploadedData: InputDataTypeLabel[][], mapping: InputToSqlMapping[], client, token: string) {
    const json = await this.transformData.transform(this.transformData.convertSpreadsheetRecords(uploadedData, mapping), client)
    return this.storeFactory.create(token, json);
  }
}
