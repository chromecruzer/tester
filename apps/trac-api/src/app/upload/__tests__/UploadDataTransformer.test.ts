import {v4 as uuidv4Mock} from 'uuid';
import * as util from "util";
import {UploadDataTransformer} from "../UploadDataTransformer";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});
jest.mock('uuid', () => ({
  v4: jest.fn(),
}));
describe('UploadDataTransformer', () => {
  let clientMock, transformer, originals, updated, mockStoreFactory;
  beforeEach(() => {
    (uuidv4Mock as jest.Mock).mockImplementation(() => 'created uuid');
    clientMock = {
      query: jest.fn(),
    };
    mockStoreFactory = {
      create: jest.fn()
    }
    clientMock.query.mockImplementation(() => Promise.resolve());
    transformer = new UploadDataTransformer(mockStoreFactory, 'schema-name', 'iol');
    originals = [
      {
        uuid: 'same',
        warehouse: 'CLR34',
        location_code: 'LOC1',
        name_code: 'nameCode1',
        address_phone: '123 White House Lane 888-112-1234',
        item: 'item1',
        description: 'I describe an item1',
        lot: 'lot1223',
        quantity: 1,
        purchase_order: 'purchased for dimplewig as item1',
        received_date: new Date('10/10/2021'),
        expire_date: new Date('10/10/2024'),
        shipped_date: new Date('10/07/2021'),
      },
      {
        uuid: 'modified',
        warehouse: 'CLR34',
        location_code: 'LOC1',
        name_code: 'nameCode1',
        address_phone: '333 Easter Bunny Lane 888-112-1234',
        item: 'item1',
        description: 'I describe an item1',
        lot: 'lot1225',
        quantity: 1,
        purchase_order: 'purchased for dimplewig as item1',
        received_date: new Date('10/10/2021'),
        expire_date: new Date('10/10/2024'),
        shipped_date: new Date('10/07/2021'),
      },
      {
        uuid: 'removed',
        warehouse: 'CLR22',
        location_code: 'LOC2',
        name_code: 'nameCode1',
        address_phone: '200 Mickey Mouse Road 888-112-1234',
        item: 'mouse',
        description: 'I describe a mouse',
        lot: 'lot1226',
        quantity: 1,
        purchase_order: 'purchased for the old lady a mouse',
        received_date: new Date('10/10/2021'),
        expire_date: new Date('10/10/2024'),
        shipped_date: new Date('10/07/2021'),
      },
    ];
    updated = [
      [
        'CLR34',
        'LOC1',
        'nameCode1',
        '123 White House Lane 888-112-1234',
        'item1',
        'I describe an item1',
        'lot1223',
        1,
        'purchased for dimplewig as item1',
        new Date('10/10/2021'),
        new Date('10/10/2024'),
        new Date('10/07/2021'),
      ],
      [
        'CLR34',
        'LOC1',
        'nameCode1',
        '333 Jack Frost Lane 888-112-1234',
        'item1',
        'I describe an item1',
        'lot1225',
        1,
        'purchased for dimplewig as item1',
        new Date('10/10/2021'),
        new Date('10/10/2024'),
        new Date('10/07/2021'),
      ],
      [
        'CLR22',
        'LOC2',
        'nameCode1',
        '200 Donald Duck Street 888-112-1234',
        'duck',
        'I describe a duck',
        'lot1227',
        1,
        'purchased for the old wolf a duck',
        new Date('10/10/2021'),
        new Date('10/10/2024'),
        new Date('10/07/2021'),
      ],
    ];
  });
  describe('#transform', () => {
    it('should return correct groupings of records', () => {
      clientMock.query.mockReturnValueOnce(Promise.resolve({rows: originals}))
        .mockReturnValueOnce(Promise.resolve({rows: [{uuid: 'removed'}]}));
      return transformer.transform(updated, clientMock, 'session-token')
        .then(() => {
          expect(mockStoreFactory.create.mock.calls).toMatchSnapshot();
        });
    });
  });
});
