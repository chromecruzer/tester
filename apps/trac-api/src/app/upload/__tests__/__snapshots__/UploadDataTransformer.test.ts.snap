// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadDataTransformer #transform should return correct groupings of records 1`] = `
Array [
  Array [
    "session-token",
    Object {
      "adds": Array [
        Object {
          "address_phone": "200 Donald Duck Street 888-112-1234",
          "description": "I describe a duck",
          "expire_date": 2024-10-10T05:00:00.000Z,
          "item": "duck",
          "location_code": "LOC2",
          "lot": "lot1227",
          "name_code": "nameCode1",
          "purchase_order": "purchased for the old wolf a duck",
          "quantity": 1,
          "received_date": 2021-10-10T05:00:00.000Z,
          "shipped_date": 2021-10-07T05:00:00.000Z,
          "warehouse": "CLR22",
        },
      ],
      "modifies": Array [
        Object {
          "new": Object {
            "address_phone": "333 Jack Frost Lane 888-112-1234",
            "description": "I describe an item1",
            "expire_date": 2024-10-10T05:00:00.000Z,
            "item": "item1",
            "location_code": "LOC1",
            "lot": "lot1225",
            "name_code": "nameCode1",
            "purchase_order": "purchased for dimplewig as item1",
            "quantity": 1,
            "received_date": 2021-10-10T05:00:00.000Z,
            "shipped_date": 2021-10-07T05:00:00.000Z,
            "uuid": "modified",
            "warehouse": "CLR34",
          },
          "originalUuid": "modified",
        },
      ],
      "removedUuids": Array [
        "removed",
      ],
    },
  ],
]
`;
